Source/DirectoryWatcher.swift

```diff
@@ -86,30 +93,45 @@ public class DirectoryWatcher {
     self.dispatchSource = nil
   }
   
-  func handleChangeEvent() {
+  func getCurrentFiles() -> [URL] {
+    do {
+      return try FileManager.default.contentsOfDirectory(
+        at: url,
+        includingPropertiesForKeys: [.creationDateKey, .typeIdentifierKey],
+        options: [.skipsHiddenFiles]
+      )
+    } catch {
+      delegate?.directoryWatcher(self, error: error)
+    }
+    return []
+  }
+  
+  func handleChangeEvent() throws {
     let currentFiles = self.currentFiles
-    let newFiles = listNewFiles(lastFiles: lastFiles, currentFiles: currentFiles)
-    let deletedFiles = listDeletedFiles(lastFiles: lastFiles, currentFiles: currentFiles)
+    let newFiles = try listNewFiles(lastFiles: lastFiles, currentFiles: currentFiles)
+    let deletedFiles = try listDeletedFiles(lastFiles: lastFiles, currentFiles: currentFiles)
     
     let changes = DirectoryChangeSet(newFiles: newFiles, deletedFiles: deletedFiles)
     delegate?.directoryWatcher(self, changed: changes)
     
     lastFiles = currentFiles
   }
   
-  func listNewFiles(lastFiles: [URL], currentFiles: [URL]) -> [URL] {
-    return createDiff(left: currentFiles, right: lastFiles)
+  func listNewFiles(lastFiles: [URL], currentFiles: [URL]) throws -> [URL] {
+    try createDiff(left: currentFiles, right: lastFiles)
   }
   
-  func listDeletedFiles(lastFiles: [URL], currentFiles: [URL]) -> [URL] {
-    return createDiff(left: lastFiles, right: currentFiles)
+  func listDeletedFiles(lastFiles: [URL], currentFiles: [URL]) throws -> [URL] {
+    try createDiff(left: lastFiles, right: currentFiles)
   }
   
-  func createDiff(left: [URL], right: [URL]) -> [URL] {
-    return Set(left).subtracting(right).sorted { (url1, url2) -> Bool in
-      let date1 = try! url1.resourceValues(forKeys: [.creationDateKey]).creationDate!
-      let date2 = try! url2.resourceValues(forKeys: [.creationDateKey]).creationDate!
-      return date1 > date2
-    }
+  func createDiff(left: [URL], right: [URL]) throws -> [URL] {
+    try Set(left).subtracting(right).sorted { try $0.creationDate() > $1.creationDate() }
+  }
+}
+
+private extension URL {
+  func creationDate() throws -> Date {
+    try resourceValues(forKeys: [.creationDateKey]).creationDate ?? .distantPast
```

Comment:
@dworak Can the usage of `.distantPast` here lead to some possible logical errors in the future? For example, if older file returns correct date whereas newer file doesn't return `creationDate` and we return `.distantPast` instead of it.

Reply:
> @dworak Can the usage of `.distantPast` here lead to some possible logical errors in the future? For example, if older file returns correct date whereas newer file doesn't return `creationDate` and we return `.distantPast` instead of it.

UPDATE -> creationDate is used for sorting reasons only so using `.distantPast` will not allow loosing any file.